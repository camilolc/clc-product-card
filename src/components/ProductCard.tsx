import React, { createContext, CSSProperties } from "react";
import styles from "../styles/styles.module.css";

import { useProduct } from "../hooks/useProduct";

import {
  Product,
  ProductContextProps,
  onChangeArgs,
  InitialValues,
  ProductCardHandlers
} from "../interfaces/interfaces";

//Interface
export interface Props {
  product: Product;
  // children: ReactElement | ReactElement[];
  children: (args: ProductCardHandlers)=> JSX.Element;
  className?: string;
  style?: CSSProperties;
  onChange?: (args: onChangeArgs) => void;
  value?: number;
  initialValues?: InitialValues;
}
//Context
export const ProductContext = createContext({} as ProductContextProps);
const { Provider } = ProductContext;

export const ProductCard = ({
  children,
  product,
  className,
  style,
  onChange,
  value,
  initialValues
}: Props) => {
  const { counter, increaseBy, maxCount,isMaxCountReached, reset } = useProduct({ onChange, product,value, initialValues });
  return (
    <Provider
      value={{
        counter,
        increaseBy,
        product,
        maxCount
      }}
    >
      <div className={`${styles.productCard} ${className}`} style={style}>
        {
        children({
          count: counter,
          maxCount: initialValues?.maxCount,
          isMaxCountReached,
          product,
          
          increaseBy,
          reset
        }
        )}       
      </div>
    </Provider>
  );
};
