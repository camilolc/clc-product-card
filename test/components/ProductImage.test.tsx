

import React from "react";

import renderer from 'react-test-renderer';

import {ProductCard, ProductImage} from "../../src/components";
import { product2 } from "../data/products";


describe('ProductImage',()=>{

    test('Debe mostar la imagen personalizada',()=>{

        const wrapper = renderer.create(
            <ProductImage img='test.com' />
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })


    test('Debe mostrar el componente con la imagen del producto',()=>{

        const wrapper = renderer.create(
            <ProductCard product={product2}>
                {
                    ()=>(
                        <ProductImage />
                    )
                }

            </ProductCard>
        )

        expect(wrapper.toJSON()).toMatchSnapshot();
    })
})