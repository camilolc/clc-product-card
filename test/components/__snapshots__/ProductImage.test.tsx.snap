// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductImage Debe mostar la imagen personalizada 1`] = `
<img
  className="productImg undefined"
  src="test.com"
/>
`;

exports[`ProductImage Debe mostrar el componente con la imagen del producto 1`] = `
<div
  className="productCard undefined"
>
  <img
    className="productImg undefined"
    src="./coffee-mug.png"
  />
</div>
`;
